pkg/manager/internal.go

```diff
@@ -532,19 +559,35 @@ func (cm *controllerManager) waitForCache() {
 	if cm.startCache == nil {
 		cm.startCache = cm.cache.Start
 	}
-	go func() {
-		if err := cm.startCache(cm.internalStop); err != nil {
-			cm.errSignal.SignalError(err)
-		}
-	}()
+	cm.startRunnable(RunnableFunc(func(stop <-chan struct{}) error {
+		return cm.startCache(stop)
+	}))
 
 	// Wait for the caches to sync.
 	// TODO(community): Check the return value and write a test
 	cm.cache.WaitForCacheSync(cm.internalStop)
+	// TODO: This should be the return value of cm.cache.WaitForCacheSync but we abuse
+	// cm.started as check if we already started the cache so it must always become true.
+	// Making sure that the cache doesn't get started twice is needed to not get a "close
+	// of closed channel" panic
 	cm.started = true
 }
 
 func (cm *controllerManager) startLeaderElection() (err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	cm.mu.Lock()
+	cm.leaderElectionCancel = cancel
+	cm.mu.Unlock()
+
+	if cm.onStoppedLeading == nil {
+		cm.onStoppedLeading = func() {
+			// We have to exit here, otherwise the graceful shutdown or anything
+			// else that keeps the binary running might allow controllers that
+			// need leader election to run after we lost the lease.
+			log.Info("leader lease lost, exiting")
+			os.Exit(0)
```

Comment:
raise a panic in case we've got defers and other stuff folks want to finish?  Nice for cleanup of temporary files and such, esp when testing

Reply:
The problem is that entering this is part of a normal shutdown of a `Manager` that has leader election enabled, so panicing here would mean that all controller-managers that have LE enabled will panic on shutdown. For tests, the whole func can be overriden.

I am not really sure what best to do here. The `os.Exit` is not nice and we also always exit with the same code, because there are different scenarios this has to cover for but which we can not distinguish:
* We lost the leader lock unexpectedly
* We just ended normally without any error
* We had an error during graceful shutdown (we will just swallow that because its returned from the `manager` but by the time that happens we will already have exited the binary here)

Reply:
Maybe push the problem down by documenting "You must exit the binary immediately after the manager ended when leader election is enabled" (and removing the os.Exit())?

Reply:
Alternatively provide a way for external users to configure this, and default it to `os.Exit`?  The immediate `Exit` is just annoying when dealing with stuff like defers.

Reply:
Okay so after thinking about this some more, I still think what we do today is not correct and unsafe, but its not trivial to improve this so I would like to have a dedicated discussion and PR for that.

I have updated this PR to disable gracefulShutdown when the leader election is lost.